File: src/main/model/problem.js
const util = require("../util");
const Model = require("./model").Model;

class Datum {
    static async construct(id, num) {
        var d = new Datum();
        d.input = await util.db.getKey(`/problems/${id}/input/in${num}.txt`);
        d.answer = await util.db.getKey(`/problems/${id}/output/out${num}.txt`);
        return d;
    }
    toJSON() {
        return {
            input: this.input,
            output: this.output
        }
    }
}

class Problem extends Model {
    static get detailsFile() { return "/problems/{}/problem.json"; }
    get detailsFile() { return "/problems/{}/problem.json"; }
    static get fields() { return ["id", "title", "description", "statement", "input", "output", "constraints", "samples", "tests"]; }
    get fields() { return ["id", "title", "description", "statement", "input", "output", "constraints", "samples", "tests"]; }
    static get folder() { return "/problems"; }
    async getSampleData() {
        let samples = [];
        for (var i = 0; i < this.samples; i ++) {
            samples.push(await Datum.construct(this.id, i));
        }
        return samples;
    }
    async getTestData() {
        let tests = [];
        for (var i = 0; i < this.tests; i ++) {
            tests.push(await Datum.construct(this.id, i));
        }
        return tests;
    }
    async toJSON() {
        let json = super.toJSON();
        const sampleData = await this.getSampleData();
        json.sampleData = sampleData.map(datum => datum.toJSON());
        return json;
    }
};

exports.Problem = Problem;

File: src/main/util/db/problem.py
from code.util.db import getKey, setKey, listSubKeys, deleteKey
from uuid import uuid4

problems = {}

class Datum:
    def __init__(self, input, output):
        self.input = input
        self.output = output
    
    def get(id: str, num: int):
        input = getKey(f"/problems/{id}/input/in{num}.txt")
        output = getKey(f"/problems/{id}/output/out{num}.txt")
        return Datum(input, output)
    
    def toJSON(self):
        return {
            "input": self.input,
            "output": self.output
        }

class Problem:
    saveCallbacks = []
    def __init__(self, id=None):
        if id != None:
            details = getKey(f"/problems/{id}/problem.json")
            self.id          = details["id"]
            self.title       = details["title"]
            self.description = details["description"]
            self.statement   = details["statement"]
            self.input       = details["input"]
            self.output      = details["output"]
            self.constraints = details["constraints"]
            self.samples     = int(details["samples"])
            self.tests       = int(details["tests"])
            self.sampleData  = [Datum.get(id, i) for i in range(self.samples)]
            self.testData    = [Datum.get(id, i) for i in range(self.tests)]
        else:
            self.id          = None
            self.title       = None
            self.description = None
            self.statement   = None
            self.input       = None
            self.output      = None
            self.constraints = None
            self.samples     = 0
            self.tests       = 0
            self.sampleData  = []
            self.testData    = []

    def get(id: str):
        if id in problems:
            return problems[id]
        return None
    
    def toJSONSimple(self):
        return {
            "id":          self.id,
            "title":       self.title,
            "description": self.description,
            "statement":   self.statement,
            "input":       self.input,
            "output":      self.output,
            "constraints": self.constraints,
            "samples":     self.samples,
            "tests":       self.tests,
        }

    def save(self):
        if self.id == None:
            self.id = str(uuid4())
            problems[self.id] = self
        setKey(f"/problems/{self.id}/problem.json", self.toJSONSimple())
        self.sampleData  = [Datum.get(id, i) for i in range(self.samples)]
        for i, datum in enumerate(self.testData):
            setKey(f"/problems/{self.id}/input/in{i}.txt", datum.input)
            setKey(f"/problems/{self.id}/output/out{i}.txt", datum.output)
        for callback in Problem.saveCallbacks:
            callback(self)
    
    def delete(self):
        deleteKey(f"/problems/{self.id}")
        del problems[self.id]
        
    def toJSON(self):
        json = self.toJSONSimple()
        json.sampleData = [datum.toJSON() for datum in self.sampleData]
        return json
    
    def toJSONFull(self):
        json = self.toJSONSimple()
        json["testData"] = [datum.toJSON() for datum in self.testData]
        return json
    
    def allJSON():
        return [problems[id].toJSONSimple() for id in problems]
    
    def forEach(callback: callable):
        for id in problems:
            callback(problems[id])
    
    def onSave(callback: callable):
        Problem.saveCallbacks.append(callback)
    
    def all():
        return [problems[id] for id in problems]

for id in listSubKeys("/problems"):
    problems[id] = Problem(id)
